app.go

```diff
@@ -449,22 +456,46 @@ func ValidateApp(opts ...Option) error {
 	return app.Err()
 }
 
+// logBuffer will buffer all messages until a logger has been
+// initialized.
+type logBuffer struct {
+	entries []fxlog.Entry
+	logger  fxlog.Logger
+}
+
+// Log buffers or logs an entry.
+func (l *logBuffer) Log(entry fxlog.Entry) {
+	if l.logger == nil {
+		l.entries = append(l.entries, entry)
+	} else {
+		l.logger.Log(entry)
+	}
+}
+
+// Connect flushes out all buffered entries to a logger and resets them.
+func (l *logBuffer) Connect(logger fxlog.Logger) {
+	l.logger = logger
+	for _, e := range l.entries {
+		logger.Log(e)
+	}
+	l.entries = []fxlog.Entry{}
+}
+
 // New creates and initializes an App, immediately executing any functions
 // registered via Invoke options. See the documentation of the App struct for
 // details on the application's initialization, startup, and shutdown logic.
 func New(opts ...Option) *App {
-	logger := fxlog.DefaultLogger(os.Stderr)
+	logBufferer := &logBuffer{}
 
 	app := &App{
-		log:          logger,
+		logBuffer:    logBufferer,
```

Comment:
nit: just logBuffer as the name should suffice. Although in this case, we don't actually need the variable.

```suggestion
	app := &App{
		logBuffer:    new(logBuffer),
```